pkg/shell/env.go

```diff
@@ -0,0 +1,145 @@
+package shell
+
+import (
+	"fmt"
+	"io/ioutil"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/semaphoreci/agent/pkg/api"
+	"github.com/semaphoreci/agent/pkg/config"
+)
+
+type Environment struct {
+	env map[string]string
```

Comment:
Let's double-check if it is ok to keep this as a map to not introduce anything unexpected. One assumption that is broken here is that the order of the env vars is no longer guaranteed.

Does this matter?

It could while adding secrets. Let's take this example:

``` yaml
secret A:
  TOKEN=abc

secret B:
  TOKEN=foo
```

and a pipeline:

``` yaml
job:
  secrets:
    - name: A
    - name: B
```

Now, if we store the values in a key-value array `[{TOKEN=abc}, {TOKEN=foo}]` it is guaranteed that every time we run a job, the value of the TOKEN will be `foo`.

If we use maps, this guarantee is no longer true.

Reply:
Now that I re-read it, it seems that I've missed that we are adding the entries to the map "in-order" they are defined.

Which makes my previous statement not true.
Either way, it is good to double-check.

Reply:
The env var names are ordered in `Append()` and `ToFile()` to help with testing, since it's hard to assert the output if we don't guarantee order.

However, those methods are only called after the environment is already created with `CreateEnvironment()`.
The `CreateEnvironment()` method is the one that takes the `[]api.EnvVar`s and the `[]config.HostEnvVar`s and puts them into a map, one by one, respecting the order of those slices. So, if a env var name is duplicated, its value will be the value of the last env var from the host (if any), or the value of the last env var in the job request.

I'll add a test for it.